sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -424,8 +424,18 @@ case class SortMergeJoinExec(
     // A list to hold all matched rows from buffered side.
     val clsName = classOf[ExternalAppendOnlyUnsafeRowArray].getName
 
+    // Flag to only buffer first matched row, to avoid buffering unnecessary rows.
+    val onlyBufferFirstMatchedRow = (joinType, condition) match {
+      case (LeftSemi, None) => true
+      case _ => false
+    }
+    val inMemoryThreshold =
+      if (onlyBufferFirstMatchedRow) {
```

Comment:
How about moving this branch into the `getInMemoryThreshold` side?
```

  // Flag to only buffer first matched row, to avoid buffering unnecessary rows.
  private lazy val onlyBufferFirstMatchedRow = (joinType, condition) match {
    case (LeftSemi, None) => true
    case _ => false
  }

  private def getInMemoryThreshold: Int = {
    if (onlyBufferFirstMatchedRow) {
      1
    } else {
      sqlContext.conf.sortMergeJoinExecBufferInMemoryThreshold
    }
  }
```

Reply:
+1, `lazy val` can probably be `def` as the logic is super simple

Reply:
Good call. Actually the non-code-gen path can also depend on this, so I make it just a `val` now.